### PR TITLE
Finite precision arithmetic coder implementation

### PR DESCRIPTION
**Pull request by kedartatwawadi**
NOTE: converting to draft PR, as it is not ready to be merged yet. But feel free to review it while I make fixes

NOTE: The implementation is finally working. But a few things remain, which I will try to add in a day or two. 
- [ ] Detailed algorithm writeup/documentation
- [ ] This implementation will probably fail tests based on #32 . I am planning to add this hopefully in a day
NOTE: things work fine now!

### CODE REVIEW

File: compressors/arithmetic_coding.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+
+class ArithmeticEncoder(DataEncoder):
+    """Finite precision Arithmetic encoder
+
+    The encoder are decoders are based on the following sources:
+    - https://youtu.be/ouYV3rBtrTI: This series of videos on Arithmetic coding are a very gradual but a great
+    way to understand them
+    - Charles Bloom's blog: https://www.cbloom.com/algs/statisti.html#A5
+    - There is of course the original paper: ADD LINK
```
Comment (shubhamchandak94) [inline compressors/arithmetic_coding.py:17]: https://dl.acm.org/doi/abs/10.1145/214762.214771
link below doesn't need access
https://web.stanford.edu/class/ee398a/handouts/papers/WittenACM87ArithmCoding.pdf

File: compressors/arithmetic_coding.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        self.freqs = freqs
+
+        # params
+        self.DATA_BLOCK_SIZE_BITS = 32  # represents the size of the data block
+        self.PRECISION = precision
+        self.FULL = 1 << (precision)
+        self.HALF = 1 << (precision - 1)
+        self.QTR = 1 << (precision - 2)
+
+    def shrink_range(self, s: Any, low: int, high: int) -> Tuple[int, int]:
```
Comment (shubhamchandak94) [inline compressors/arithmetic_coding.py:143]: is this same as the encoder function? If yes, can combine into one by creating a helper class

**Review by shubhamchandak94 — COMMENTED** [review]
Thanks, took a quick look but I think more comments will be helpful to understand what's going on

**Comment by shubhamchandak94** [comment]
why is SFE part of this PR also?

File: compressors/arithmetic_coding.py (hunk 1 above)
Comment (kedartatwawadi) [reply compressors/arithmetic_coding.py:17]: Thanks! ll add

File: compressors/arithmetic_coding.py (hunk 2 above)
Comment (kedartatwawadi) [reply compressors/arithmetic_coding.py:143]: yes, there is lot of overlap. I will try to clean this up later today

**Comment by kedartatwawadi** [comment]
@shubhamchandak94 : SFE diff will go away once we merge that PR.. I used some utility functions from that branch, so based this branch on that

**Comment by kedartatwawadi** [comment]
@shubhamchandak94 : I will add more comments/ links to the arithmetic coder write-up to explain what is happening as it will be difficult to explain everything in comments

File: compressors/arithmetic_coding.py (hunk 2 above)
Comment (kedartatwawadi) [reply compressors/arithmetic_coding.py]: Done

File: compressors/arithmetic_coding.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+                low = (low - self.params.QTR) << 1
+                high = (high - self.params.QTR) << 1
+                state = (state - self.params.QTR) << 1
+
+                if num_bits_consumed < arith_bitarray_size:
+                    bit = encoded_bitarray[num_bits_consumed]
+                    state += bit
+                num_bits_consumed += 1
+
+        # # NOTE: we might have loaded in additional bits not added by the arithmetic encoder
```
Comment (shubhamchandak94) [inline compressors/arithmetic_coding.py:347]: @kedartatwawadi - I think the understand the whole thing except for this last part. If you can explain it here or in the wiki, I'm happy to approve this.

**Review by shubhamchandak94 — COMMENTED** [review]
Left a comment

File: core/prob_dist.py [hunk 4]
```diff
@@ -187,13 +187,11 @@ def freq_list(self):
         return [self.freq_dict[s] for s in self.alphabet]
 
     @property
-    @cache
```
Comment (tpulkit) [inline core/prob_dist.py:190]: Can you remind me the issue with adding this cache and why we decided to remove it?

File: compressors/arithmetic_coding.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+        num_bits_consumed += self.params.DATA_BLOCK_SIZE_BITS
+
+        return DataBlock(decoded_data_list), num_bits_consumed
+
+
+def test_bitarray_for_specific_input():
+    """
+    manually perform the encoding and check if the bitarray matches
+    """
+    # TODO: Kedar
```
Comment (tpulkit) [inline compressors/arithmetic_coding.py:367]: @Kedar: are you planning to implement it before accepting PR?

File: compressors/arithmetic_coding.py [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+                self.freq_model.freqs_current.total_freq < self.params.MAX_ALLOWED_TOTAL_FREQ
+            ), err_msg
+
+            # shrink range
+            # i.e. the core Arithmetic encoding step
+            low, high = ArithmeticEncoder.shrink_range(self.freq_model.freqs_current, s, low, high)
+            ## update the freq model for encoding the next symbol
+            self.freq_model.update_model(s)
+
+            # perform re-normalizing range
```
Comment (tpulkit) [inline compressors/arithmetic_coding.py:193]: @kedartatwawadi: More comments can be added in this section unless you were thinking of explaining all this in detail in write-up. Given just the code, it's hard to decode the logic based on just the comments. I am not sure what should be our level of intention here with the comments.

Comment (kedartatwawadi) [reply compressors/arithmetic_coding.py:194]: ll do

File: compressors/arithmetic_coding.py (hunk 5 above)
Comment (kedartatwawadi) [reply compressors/arithmetic_coding.py:368]: yes ll add

**Comment by kedartatwawadi** [comment]
@tpulkit  -> ll address some of the tests issues later. Lets merge the PR. Can you please approve?

**Pull request merged**